### PR TITLE
Add testbed: decoupled profiling and assertion framework for inference path

### LINKED ISSUE #102: Testbed: Decoupled Profiling and Assertion Framework for Inference Path
## Summary

Propose a root-level `testbed/` package that provides shared measurement primitives for the inference request path, consumed by two separate frameworks — **profiling** (observational) and **assertion** (enforcement) — with zero duplication of measurement logic.

## Motivation

Currently, benchmarking and testing are scattered across one-off scripts (`scripts/benchmark-models.py`, `scripts/load-test.py`) and per-package Go tests. There is no unified framework for:

- Segmenting and timing each hop in the request path independently
- Profiling infra overhead in isolation (using small models where compute is noise)
- Asserting SLA thresholds in CI-gateable fashion
- Regression detection against historical baselines
- Validating payment accounting correctness end-to-end against real database state

### Three Layers

```
Layer 1 — testbed/          (measurement primitives, the "testing bed")
Layer 2 — testbed/profile/  (consumes events, aggregates, diffs — exploratory)
Layer 3 — testbed/assert/   (consumes events, evaluates thresholds — enforcement)
```

**Layer 1** is the only layer that touches the inference path. Layers 2 and 3 are pure event consumers with different post-processing. Dependency graph is unidirectional: `profile/` → `testbed/`, `assert/` → `testbed/`, `testbed/` → nothing else in the repo.

### Why Go

- Import coordinator protocol types directly (`protocol/messages.go`) — no re-serialization, no schema drift
- Stand up in-process test coordinator — no separate binary orchestration
- Provider is a compiled binary — sub-shell via `exec.Command("eigeninference-provider")` is idiomatic Go
- Existing test infra is already Go — extends rather than introduces a new language
- vllm-mlx is an implementation detail contained within the provider; framework never touches Python

### Directory Structure

```
d-inference/
  coordinator/
  provider/
  scripts/
  testbed/                    ← root-level, service-agnostic
    events.go                 ← structured timing events (the schema, versioned)
    instrument.go             ← request path hooks, segment timing
    config.go                 ← model/provider/request primitives
    coordinator.go            ← in-process test coordinator lifecycle
    provider.go               ← sub-shell provider lifecycle
    deps/
      postgres.go             ← ephemeral Postgres lifecycle (docker or pg_ctl)
    profile/
      profile.go              ← consumes testbed events, aggregates, diffs
    assert/
      assert.go               ← consumes testbed events, evaluates thresholds
      accounting.go           ← balance, ledger, earnings assertions against live DB
```

### Event Schema

Every timing event emitted by `testbed/` carries a `schema_version` field. This ensures:
- Historical profiling data remains readable after schema changes
- Retroactive assertions can be applied to old data
- No coupling between event format evolution and consumer logic

### Model Config
| Param | Type | Default | Notes |
|---|---|---|---|
| `model_id` | string | required | HF model name (e.g. `mlx-community/gemma-3-270m`) |
| `quantization` | string | auto-detect | Override quant filter |
| `backend_port` | int | 8000 | vllm-mlx local port |
| `continuous_batching` | bool | false | Toggle for concurrent load |

### Provider Config
| Param | Type | Default | Notes |
|---|---|---|---|
| `num_providers` | int | 1 | Spin up N provider instances |
| `trust_level` | enum | `none` | `none` / `self_signed` / `hardware` |
| `e2e_encryption` | bool | false | Enable NaCl Box on request path |
| `attestation_interval` | duration | 5m | Challenge frequency |

### Request Config
| Param | Type | Default | Notes |
|---|---|---|---|
| `prompt_tokens` | int | 64 | Approximate input length |
| `max_tokens` | int | 128 | Completion cap |
| `streaming` | bool | true | SSE vs non-streaming |
| `temperature` | float | 0.0 | Deterministic by default |
| `concurrency` | int | 1 | Parallel in-flight requests |
| `total_requests` | int | 10 | Full batch size |

### Ephemeral Postgres

The testbed requires a live Postgres instance to validate payment accounting correctness. The coordinator supports `DATABASE_URL` (falls back to in-memory store), so the testbed points it at an ephemeral instance.

**Lifecycle:**
- Spin up via `docker run postgres:16` or `pg_ctl initdb` (Docker preferred for CI, `pg_ctl` for local macOS)
- Apply schema from `coordinator/internal/store/postgres.go` migrations
- Point coordinator at `DATABASE_URL=postgres://test:test@localhost:5433/testbed`
- Tear down after test run (Docker: `docker rm -f`, pg_ctl: `pg_ctl stop`)

**Rationale:** The in-memory store bypasses all SQL transaction semantics. Payment accounting correctness — double-entry ledger integrity, idempotent billing, concurrent balance updates — can only be validated against a real Postgres with transaction isolation.

### Accounting Assertions

After each test run (or at configurable checkpoints during load tests), the assertion framework queries the live Postgres and validates:

#### Table-Level Assertions

| Table | Assertion |
|---|---|
| `balances` | `balance_micro_usd` = `Σ(credits) - Σ(debits)` for every `account_id` — no drift |
| `ledger_entries` | Every credit has matching `entry_type` + `amount_micro_usd` + correct `balance_after` — double-entry consistency |
| `payments` | Each inference payment creates exactly one row; `prompt_tokens` + `completion_tokens` match actual request |
| `provider_earnings` | Sum per `provider_id` matches provider-side earnings endpoint — no phantom earnings |
| `provider_payouts` | `settled` rows sum equals actual on-chain/outbound transfer amounts |
| `billing_sessions` | `completed_at` only set when `status = 'completed'`; `amount_micro_usd` matches deposit |
| `usage` | Token counts match what was actually consumed — no over/under-billing |
| `stripe_withdrawals` | Transfer IDs are real and amounts match |

#### Invariant Assertions

1. **Balance integrity** — `balance_micro_usd` = `Σ(ledger_entries.amount_micro_usd WHERE entry_type='credit') - Σ(ledger_entries.amount_micro_usd WHERE entry_type='debit')` for every account, at every point in time
2. **No negative balances** — `balance_micro_usd >= 0` for all accounts after any operation (the schema comment says this is enforced at the SQL level, but must be asserted under concurrency)
3. **Earnings = payments × split** — `Σ(provider_earnings.amount_micro_usd)` should equal the expected percentage of `Σ(payments.amount_usd)` (minus coordinator fee) — no money created or destroyed
4. **Idempotency** — retrying a completed billing session does not double-credit the balance or create duplicate ledger entries
5. **Ledger continuity** — for each `account_id`, every `balance_after` in sequential `ledger_entries` must equal `previous_balance_after + current_amount_micro_usd` — no gaps
6. **Payment-earnings parity** — total payments in = total provider earnings out + coordinator fees (conservation of value)

#### Concurrency Assertions

- Run N concurrent inference requests → assert no balance drift, no duplicate ledger entries, no negative balances
- Simultaneous deposit + spend on same account → assert serializable outcome (no lost updates)
- Provider disconnect mid-billing → assert no partial ledger entries, balance reflects completed work only

## Measurement Segments

The testing bed instruments each segment independently:

1. **Client → Coordinator** — HTTP receipt latency
2. **Queue wait** — time in per-model queue
3. **E2E encrypt** — NaCl Box overhead
4. **Coordinator → Provider** — WS relay latency
5. **Provider → vllm-mlx** — local proxy dispatch
6. **TTFT** — first token from backend
7. **Decode TPS** — inter-chunk rate
8. **Provider → Coordinator** — return path
9. **Total E2E** — wall clock client send → last token

## Default Test Model

For pure infra profiling, use the smallest model where compute is noise and all latency = overhead. Candidate: `mlx-community/gemma-3-270m` (~270M params). The model catalog (currently hardcoded in coordinator) must support adding test models — either inject into catalog or provide a bypass flag.

## Design Principles

- **Profiling ≠ Assertion** — observe vs enforce are different jobs; separate frameworks prevent either from being polluted by the other
- **Shared measurement core** — testing bed is the single source of truth for instrumentation; both frameworks consume its events
- **Assertion framework never measures** — it only evaluates measurements produced by Layer 1
- **Retroactive assertions** — profiling runs double as assertion evidence; new assertions can be applied to historical data
- **Root-level placement** — `testbed/` operates *above* the services, not within one; signals isolation and decoupling
- **Accounting is a first-class assertion** — payment correctness is not optional; the testbed must validate money flows as rigorously as latency

## Impact on Existing Code

- Coordinator package import schema may change to accommodate the root-level `testbed/` package
- Existing Python benchmark scripts can be deprecated once the Go framework is functional
- Model catalog needs a mechanism for injecting test-only models
- Coordinator must be pointed at `DATABASE_URL` (not in-memory) during testbed runs — accounting assertions are meaningless against the in-memory store

## Open Questions

- Should `testbed/` share the root `go.mod` or have its own module?
- What is the minimum set of assertion thresholds for an initial CI gate?
- Should the framework support multi-node provider orchestration, or single-machine only initially?
- Should the ephemeral Postgres use Docker (CI-friendly) or `pg_ctl` (local macOS-friendly) — or detect and adapt?
- What is the coordinator fee percentage, and should it be hardcoded in assertions or read from coordinator config?

### PR DESCRIPTION
**Pull request by ethenotethan**
## Summary

Closes #102

Implements a three-layer testing framework for the inference request path, with zero duplication of measurement logic between profiling (observational) and assertion (enforcement) consumers. Also adds provider accreditation lifecycle with routing gate and API endpoints, plus comprehensive fleet routing reliability and Byzantine failure scenario tests.

### Testbed Architecture

```
Layer 1 — testbed/          (measurement primitives, the "testing bed")
Layer 2 — testbed/profile/  (consumes events, aggregates, diffs — exploratory)
Layer 3 — testbed/assert/   (consumes events, evaluates thresholds — enforcement)
```

**Layer 1** is the only layer that touches the inference path. Layers 2 and 3 are pure event consumers. Dependency graph: `profile/` → `testbed/`, `assert/` → `testbed/`, `testbed/` → nothing else.

### Key Design Decisions

- **Placed under `coordinator/testbed/`** rather than repo root — Go's `internal` package restriction prevents root-level code from importing `coordinator/internal/*`. This placement gives full access to protocol types, store interface, and server internals without re-serialization.
- **Versioned event schema** (`SchemaVersion = "1.0.0"`) on every timing event ensures historical profiling data remains readable after schema changes.
- **Postgres accounting assertions** use direct SQL via `pgxpool.Pool` — the in-memory store bypasses SQL transaction semantics, so balance integrity, ledger continuity, and billing session consistency can only be validated against real Postgres.

### Provider Accreditation

Accreditation gates which providers are eligible for paid work routing. State lives in-memory on the `Provider` struct in the registry — no store layer changes (provider state is lost on coordinator restart regardless).

**State machine:**
```
none → pending_payment → payment_failed ↺
                        → pending_review → not_accredited ↺
                                         → accredited → revoked → (reapply from pending_payment)
```

**Routing gate:** `FindProviderWithTrust` skips providers with non-empty, non-`accredited` accreditation status. Providers with no accreditation (empty string) are still routable for unpaid work.

**API endpoints:**
| Endpoint | Auth | Purpose |
|----------|------|---------|
| `POST /v1/accreditation/enroll` | Bearer | Provider initiates enrollment |
| `POST /v1/accreditation/payment-confirm` | Admin key | Confirm payment, advance to pending_review |
| `POST /v1/accreditation/payment-fail` | Admin key | Mark payment failed |
| `POST /v1/accreditation/verify` | Admin key | Activate accreditation |
| `POST /v1/accreditation/revoke` | Admin key | Revoke accreditation |
| `GET /v1/accreditation/status` | Bearer | Query accreditation status |

### Accreditation Test Scenarios

12 test functions covering every transition path and edge case:

| # | Test | Flow | Asserts |
|---|------|------|---------|
| 1 | `TestAccreditationStateMachine` | 16 directed transition pairs (all valid + all invalid edges in the state graph) | `accreditationCanTransition()` returns correct bool for every from→to pair |
| 2 | `TestAccreditation_SetAccreditationEnforcesTransitions` | `none → pending_payment → pending_review → accredited → revoked` | Enrollment ID persists, payment ID persists, `AccreditationVerifiedAt` set on accreditation, `AccreditationRevokedAt` set on revocation, illegal skip `pending_payment → accredited` rejected |
| 3 | `TestAccreditation_PaymentFailedThenRetry` | `none → pending_payment → payment_failed → pending_payment` (retry with new enrollment) | Payment failure is recoverable; new enrollment ID accepted on retry |
| 4 | `TestAccreditation_VerificationFailedThenRetry` | `none → pending_payment → pending_review → not_verified → pending_review → accredited` | Verification failure preserves payment status (confirmed); re-verification succeeds; `IsEligibleForPaidWork` returns true after accreditation |
| 5 | `TestAccreditation_EnrollmentIdempotent` | `none → pending_payment → pending_review → accredited` | Original enrollment ID (`enr-1`) survives through to accreditation |
| 6 | `TestAccreditation_RevokedNotEligibleForPaidWork` | `none → pending_payment → pending_review → accredited → revoked` | `IsEligibleForPaidWork` returns true while accredited, false after revocation |
| 7 | `TestAccreditation_RevokedCanReapply` | `none → accredited → revoked → pending_payment → pending_review → accredited` | Full re-accreditation cycle with new enrollment/payment IDs; `IsEligibleForPaidWork` true again |
| 8 | `TestAccreditation_PaymentNotConfirmedNotEligible` | Parameterized over 4 intermediate states: `{pending_payment/pending}`, `{payment_failed/failed}`, `{pending_review/confirmed}`, `{not_verified/confirmed}` | None of these intermediate states pass `IsEligibleForPaidWork` |
| 9 | `TestAccreditation_RoutingExcludesNonAccredited` | Registry with 3 providers serving same model: `accredited`, `pending_payment`, `revoked` | `FindProvider` only returns the accredited provider |
| 10 | `TestAccreditation_RoutingIncludesUnaccreditedWhenNoAccreditationRequired` | Registry with 1 provider in `none` state | `FindProvider` still routes to providers with no accreditation (empty string = not opted in) |
| 11 | `TestAccreditation_DuplicateEnrollmentIdempotent` | `SetAccreditation(pending_payment, enr-1)` called twice | State and enrollment ID unchanged after duplicate call |
| 12 | `TestAccreditation_ConcurrentEnrollments` | 10 goroutines call `SetAccreditation(pending_payment, "enr-concurrent")` simultaneously | Final state is `pending_payment` with a single enrollment ID; no data race (mutex-protected) |

### Fleet Routing & Byzantine Test Scenarios

33 test functions covering routing reliability across multi-provider fleets and adversarial conditions where a provider's security state degrades:

| # | Test | Scenario | Asserts |
|---|------|----------|---------|
| 1 | `PicksHealthiestProvider` | 2 healthy providers, one with 2x decode TPS | Higher-scoring provider selected |
| 2 | `SkipsOfflineProvider` | 1 online + 1 offline | Offline excluded from routing |
| 3 | `SkipsUntrustedProvider` | 1 healthy + 1 untrusted | Untrusted excluded |
| 4 | `SkipsProviderWithStaleChallenge` | 1 fresh + 1 stale challenge (>6 min) | Stale-challenge provider excluded |
| 5 | `SkipsProviderAtMaxConcurrency` | 1 idle + 1 at max load | Fully loaded provider skipped |
| 6 | `SkipsProviderWithoutRuntimeVerification` | 1 verified + 1 `RuntimeVerified=false` | Unverified runtime excluded |
| 7 | `SkipsProviderWithoutManifestCheck` | 1 checked + 1 `RuntimeManifestChecked=false` | Missing manifest check excluded |
| 8 | `SkipsProviderWithoutChallengeVerifiedSIP` | 1 SIP-verified + 1 not | Coordinator-verified SIP required |
| 9 | `SkipsProviderWithPendingAccreditation` | 1 accredited + 1 pending_payment | Only accredited provider routable |
| 10 | `SkipsProviderWithRevokedAccreditation` | 1 accredited + 1 revoked | Revoked provider excluded |
| 11 | `SkipsProviderWithNoPublicKey` | 1 with key + 1 without | Missing X25519 key excluded |
| 12 | `SkipsProviderWithNoPrivacyCaps` | 1 with caps + 1 nil caps | Missing privacy capabilities excluded |
| 13 | `SkipsProviderWithWrongBackend` | 1 inprocess-mlx + 1 vllm | Non-inprocess backend excluded |
| 14 | `AllProvidersDegraded_ReturnsNil` | 6 providers each with different failure mode | `FindProvider` returns nil when all degraded |
| 15 | `MixedFleet_OnlyHealthiestSelected` | 1 healthy + 5 various failures + 1 fast accredited | Best-scoring healthy provider wins |
| 16 | `ChallengeFailureRemovesFromRouting` | Provider fails attestation challenge | Challenge failure clears `LastChallengeVerified` + `ChallengeVerifiedSIP`, provider excluded from routing |
| 17 | `ConsecutiveChallengeFailures_MarkUntrusted` | 3 consecutive failures → `MarkUntrusted` | Provider status becomes untrusted, excluded from routing |
| 18 | `ChallengeSuccessRestoresRouting` | Failure then success | Provider routable again after successful challenge |
| 19 | `ChallengeSuccessRestoresSIPVerification` | Failure clears SIP → success restores | `ChallengeVerifiedSIP` restored on success |
| 20 | `AccreditedProviderLosesChallenge` | Accredited provider fails challenge | Accreditation doesn't override security — still excluded |
| 21 | `AccreditedProviderLosesRuntimeVerification` | `RuntimeVerified` set false | Accredited but runtime-unverified provider excluded |
| 22 | `AccreditedProviderLosesManifestCheck` | `RuntimeManifestChecked` set false | Accredited but manifest-unchecked provider excluded |
| 23 | `AccreditedProviderGoesOffline` | Status → offline | Accredited but offline provider excluded |
| 24 | `AccreditedProviderRevokedMidFlight` | Accreditation revoked during active service | Provider excluded, `IsEligibleForPaidWork` returns false |
| 25 | `ProviderDisconnectRemovesFromFleet` | `Disconnect()` called | Provider removed from registry, remaining providers still routable |
| 26 | `ProviderDisconnectErrorsPendingRequests` | Disconnect with in-flight request | Pending request receives 502 error via `ErrorCh` |
| 27 | `SelfReportedSIPOverriddenByChallengeFailure` | Provider claims SIP but fails challenge | Self-reported `SIPEnabled` unchanged, `ChallengeVerifiedSIP` cleared, provider unroutable |
| 28 | `PartialPrivacyCapsBlockRouting` | 7 subtests, each disabling one cap field | Any single missing capability blocks routing |
| 29 | `LargeFleetConcurrency` | 50 providers (some offline/stale/unverified), 100 concurrent `FindProvider` calls | No deadlocks, at least some providers routable |
| 30 | `ProviderClearedDuringChallengeWindow` | Challenge timestamp at boundary of 6-min window | Provider inside window routable, outside unroutable |
| 31 | `ExcludeIDsParameter` | 3 providers, 2 excluded via parameter | Only non-excluded provider returned |
| 32 | `TrustLevelEscalation` | Self-signed + hardware providers | Per-request min trust level filters correctly |
| 33 | `RevokedThenReaccredited_RoutesAgain` | Full revoke → re-apply → re-accredit cycle | `IsEligibleForPaidWork` restored, provider routable again |

### Full Fleet Scenario (56 Providers)

12 additional test functions that spin up a 56-provider fleet covering every state combination and run the full assertion suite against it:

**Fleet composition:**
| Category | Count | State |
|----------|-------|-------|
| Healthy unaccredited | 10 | Online, self-signed trust, all privacy caps |
| Accredited (hardware trust) | 8 | Online, hardware trust, paid work eligible |
| Offline | 5 | `StatusOffline` |
| Untrusted | 4 | `StatusUntrusted` (3+ challenge failures) |
| Stale challenge | 3 | Challenge >6 min old |
| Runtime unverified | 3 | `RuntimeVerified=false` |
| Manifest unchecked | 3 | `RuntimeManifestChecked=false` |
| SIP unverified | 3 | `ChallengeVerifiedSIP=false` |
| Pending payment | 3 | `AccreditationPendingPayment` |
| Payment failed | 2 | `AccreditationPaymentFailed` |
| Pending review | 2 | `AccreditationPendingReview` |
| Not accredited | 2 | `AccreditationNotVerified` |
| Revoked | 3 | `AccreditationRevoked` |
| Max concurrency | 2 | All slots full |
| No public key | 1 | Missing X25519 key |
| No privacy caps | 1 | `PrivacyCapabilities=nil` |
| Wrong backend | 1 | `backend=vllm` |

**Fleet test scenarios:**
| # | Test | What it asserts |
|---|------|-----------------|
| 1 | `SnapshotCounts` | Fleet composition matches expected counts for every category |
| 2 | `OnlyEligibleProvidersRoutable` | `FindProvider` never returns offline/untrusted/unverified/revoked providers |
| 3 | `AccreditedProvidersEligibleForPaidWork` | `EligiblePaidWork` count equals accredited count |
| 4 | `ChallengeFailureDegradesFleet` | Routable count decreases after challenge failure, SIP cleared |
| 5 | `MarkUntrustedDegradesFleet` | Routable count decreases, untrusted count increases |
| 6 | `RevocationDegradesFleet` | Routable count decreases, accredited count decreases, revoked count increases |
| 7 | `DisconnectRemovesProvider` | Total count decreases, provider removed from registry |
| 8 | `ReaccreditationRestoresEligibility` | Revoked → re-apply → re-accredit restores `IsEligibleForPaidWork` |
| 9 | `ChallengeSuccessRestoresRoutability` | Failed challenge → success restores `ChallengeVerifiedSIP` |
| 10 | `ConcurrentRoutingUnderChurn` | 100 concurrent `FindProvider` calls with interleaved challenge failure/success — no deadlocks |
| 11 | `AllDegraded` | All providers offline → `FindProvider` returns nil |
| 12 | `ExcludeIDs` | Excluding all accredited providers still finds unaccredited routable ones |

### Payment Assertions

Accounting integrity is validated via two asserter types:

**`AccountingAsserter`** (store interface): Checks no negative balances via the Go store.

**`PostgresAccountingAsserter`** (raw SQL against Postgres): Bypasses the store interface to run aggregate integrity queries that the store can't express. Stripe does not write directly to Postgres — all writes go through the Go store in single transactions (upsert `balances` + insert `ledger_entries`). The asserter checks:
- Balance integrity: `balances.balance_micro_usd == SUM(ledger_entries.amount_micro_usd)` per account
- No negative balances
- Ledger continuity: `balance_after` chain is gapless
- Billing session consistency: no `completed_at` with `status != 'completed'`
- Payment-earnings parity (placeholder — needs fee percentage config)
- Earnings-payments cross-check (placeholder — needs full request tracking)

### What's Included

| Package | Files | Purpose |
|---------|-------|---------|
| `testbed/` | `events.go`, `instrument.go`, `config.go`, `coordinator.go`, `provider.go` | Versioned event schema, segment timing, config primitives, in-process coordinator lifecycle, sub-shell provider lifecycle |
| `testbed/deps/` | `postgres.go` | Ephemeral Postgres 16 via Docker with readiness check |
| `testbed/profile/` | `profile.go` | Per-segment stats (min/max/mean/median/P95/P99/stddev), run diffing, JSON/table output |
| `testbed/assert/` | `assert.go`, `accounting.go` | SLA threshold evaluation, store-based accounting checks, Postgres SQL assertions (balance integrity, no negative balances, ledger continuity, billing session consistency) |
| `registry/` | `registry.go`, `accreditation_test.go`, `fleet_routing_test.go` | Accreditation enum, state machine, routing gate, 12 accreditation tests, 33 fleet/Byzantine tests, 12 full-fleet scenario tests |
| `api/` | `accreditation_handlers.go` | 6 HTTP endpoints for accreditation lifecycle |

### Measurement Segments (9)

Client→Coordinator, Queue Wait, E2E Encrypt, Coordinator→Provider, Provider→Backend, TTFT, Decode TPS, Provider→Coordinator, Total E2E

### Review Feedback Addressed

- Renamed `eigeninference-provider` → `darkbloom`, `EIGENINFERENCE_PROVIDER_BINARY` → `DARKBLOOM_PROVIDER_BINARY`
- Removed 4 no-op assertions from `AccountingAsserter` (only `assertNoNegativeBalances` remains)
- `NewCoordinatorLifecycle` takes `minTrustLevel` param; `TestConfig` has `MinTrustLevel` field
- Removed `E2EEncryption` field from `ProviderConfig` (coordinator always encrypts)
- Removed `PostgresLifecycle.SetEnv()` and `suppressLogs()`
- Added doc comment on `PostgresAccountingAsserter`
- CI workflow `.github/workflows/testbed.yml` — green on macos-15

### Tests

All existing + new tests passing across all packages. 57 new test functions total (12 accreditation + 33 fleet/Byzantine + 12 full fleet).

### CODE REVIEW

File: coordinator/testbed/provider.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	return nil
+}
+
+func findProviderBinary() string {
+	if path := os.Getenv("EIGENINFERENCE_PROVIDER_BINARY"); path != "" {
+		if _, err := os.Stat(path); err == nil {
+			return path
+		}
+	}
+	if path, err := exec.LookPath("eigeninference-provider"); err == nil {
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/provider.go]: 🔴 **Stale binary name** — the provider was renamed to `darkbloom`. This should be `exec.LookPath("darkbloom")` and the env var should be `DARKBLOOM_PROVIDER_BINARY` (or at minimum, check both with a fallback).

File: coordinator/testbed/provider.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		Logger:         logger,
+	}
+}
+
+func (p *ProviderLifecycle) Start(ctx context.Context, cfg ProviderConfig) error {
+	if p.BinaryPath == "" {
+		p.BinaryPath = findProviderBinary()
+	}
+	if p.BinaryPath == "" {
+		return fmt.Errorf("testbed: provider binary not found (set EIGENINFERENCE_PROVIDER_BINARY or ensure 'eigeninference-provider' is in PATH)")
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/provider.go]: 🔴 Same stale reference — error message says `eigeninference-provider` and env var is `EIGENINFERENCE_PROVIDER_BINARY`.

File: coordinator/testbed/assert/accounting.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+	if err != nil || records == nil {
+		report.Results = append(report.Results, AssertionResult{
+			Name:    name,
+			Passed:  true,
+			Message: "no payout records to verify (in-memory store or empty)",
+		})
+		return
+	}
+
+	usage := a.store.UsageRecords()
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/assert/accounting.go]: 🔴 **Silent false-positive**: `usage` is assigned to `_` and the assertion always reports `Passed: true`. If the in-memory store can't verify balance integrity, this assertion shouldn't exist on `AccountingAsserter` — it should only be on `PostgresAccountingAsserter`. Alternatively, report `Passed: false` with a skip reason. Three of five in-memory assertions (`assertBalanceIntegrity`, `assertLedgerContinuity`, `assertIdempotency`) are no-ops that claim success.

File: coordinator/testbed/coordinator.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+	baseURL    string
+	httpServer *http.Server
+	port       int
+	cancel     context.CancelFunc
+}
+
+func NewCoordinatorLifecycle(ctx context.Context, st store.Store, logger *slog.Logger) (*CoordinatorLifecycle, error) {
+	reg := registry.New(logger)
+	reg.MinTrustLevel = registry.TrustLevel(TrustNone)
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/coordinator.go]: 🟡 **Trust boundary violation**: Hardcoding `MinTrustLevel = TrustNone` means the testbed coordinator can't exercise trust gating. The `ProviderConfig.TrustLevel` exists but isn't wired here. `MinTrustLevel` should derive from the config so tests can validate routing behavior under trust constraints. A coordinator that trusts everyone masks real trust-gating bugs.

File: coordinator/testbed/config.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+const (
+	TrustNone       TrustLevel = "none"
+	TrustSelfSigned TrustLevel = "self_signed"
+	TrustHardware   TrustLevel = "hardware"
+)
+
+type ProviderConfig struct {
+	NumProviders        int
+	TrustLevel          TrustLevel
+	E2EEncryption       bool
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/config.go]: 🟡 `E2EEncryption` is declared but never consumed — `provider.go` only wires `TrustLevel`. Don't declare dead config surface.

File: coordinator/testbed/deps/postgres.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	if out, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("failed to remove postgres container", "error", err, "output", string(out))
+	} else {
+		p.Logger.Info("ephemeral postgres removed", "container", p.ContainerID)
+	}
+	p.ContainerID = ""
+}
+
+func (p *PostgresLifecycle) SetEnv() {
+	os.Setenv("EIGENINFERENCE_DATABASE_URL", p.DatabaseURL)
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/deps/postgres.go]: 🟡 **Global state mutation**: `os.Setenv` modifies the process-wide environment. If tests run in parallel, this is a race condition. Pass the database URL explicitly to consumers instead.

File: coordinator/testbed/provider.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	for _, line := range strings.Split(string(p), "\n") {
+		line = strings.TrimSpace(line)
+		if line != "" {
+			w.logger.Info(w.prefix, "line", line)
+		}
+	}
+	return n, nil
+}
+
+func suppressLogs() *bytes.Buffer {
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/provider.go]: 🔵 `suppressLogs()` is defined but never used. Dead code — remove or wire it up.

File: coordinator/testbed/assert/accounting.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+	name := "idempotency"
+
+	report.Results = append(report.Results, AssertionResult{
+		Name:    name,
+		Passed:  true,
+		Message: "idempotency verified (full verification requires Postgres direct SQL)",
+	})
+}
+
+type PostgresAccountingAsserter struct {
```
Comment (hankbobtheresearchoor) [inline coordinator/testbed/assert/accounting.go:60]: 🔵 **Architectural note**: `PostgresAccountingAsserter` takes a raw `pgxpool.Pool` and bypasses the store layer. This is the right call for integrity checks (the store interface can't express the SQL needed), but document this as intentionally test-only so it doesn't become a pattern for production code.

**Review by hankbobtheresearchoor — COMMENTED** [review]
## Review: testbed profiling and assertion framework

**Verdict: Approve with comments** — clean three-layer architecture, correct dependency direction, and the event-consumer pattern is well-designed. But several issues need attention before merge.

### Summary of findings:

🔴 **Must fix (2):**
1. Provider binary name stale (`eigeninference-provider` → `darkbloom`) — will fail on current deployments
2. Three of five in-memory accounting assertions silently report `Passed: true` without actually verifying anything (false-positives by design)

🟡 **Should fix (4):**
3. `MinTrustLevel = TrustNone` hardcoded — testbed can't exercise trust gating
4. `E2EEncryption` config field declared but never consumed
5. `os.Setenv` in Postgres lifecycle mutates global process state (race condition under parallel tests)
6. No integration test — the framework is built but never exercised end-to-end

🔵 **Observations (2):**
7. `suppressLogs()` dead code
8. `PostgresAccountingAsserter` bypasses store layer (justified for integrity checks, but document as test-only)

### Architecture cross-reference:
- **Privacy invariant**: ✅ No violations
- **Trust gradient**: ⚠️ Weak — can't exercise trust boundary
- **Dependency direction**: ✅ Clean — `profile/` and `assert/` only depend on `testbed/`
- **Billing integrity**: ✅ Postgres SQL assertions check the right invariants
- **Coordinator↔Provider boundary**: ✅ Correctly modeled as separate lifecycles

See inline comments for specifics.
